flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -885,28 +885,35 @@ class ScalarExprLowering {
     llvm_unreachable("unknown ordering");
   }
 
-  template <int KIND>
-  ExtValue genval(const Fortran::evaluate::SetLength<KIND> &x) {
-    // Change the dynamic length information without actually changing the
-    // underlying character storage.
+  // Change the dynamic length information without actually changing the
+  // underlying character storage.
+  fir::ExtendedValue
+  replaceScalarCharacterLength(const fir::ExtendedValue &scalarChar,
+                               mlir::Value newLenValue) {
     mlir::Location loc = getLoc();
-    mlir::Value newLenValue = genunbox(x.right());
-    fir::ExtendedValue lhs = gen(x.left());
-    const fir::CharBoxValue *lhsCharBox = lhs.getCharBox();
-    if (!lhsCharBox)
+    const fir::CharBoxValue *charBox = scalarChar.getCharBox();
+    if (!charBox)
       fir::emitFatalError(loc, "expected scalar character");
-    mlir::Value lhsAddr = lhsCharBox->getAddr();
+    mlir::Value charAddr = charBox->getAddr();
     auto charType =
-        fir::unwrapPassByRefType(lhsAddr.getType()).cast<fir::CharacterType>();
+        fir::unwrapPassByRefType(charAddr.getType()).cast<fir::CharacterType>();
     if (charType.hasConstantLen()) {
       // Erase previous constant length from the base type.
       fir::CharacterType::LenType newLen = fir::CharacterType::unknownLen();
```

Comment:
Is the LEN always guaranteed to be non-constant? Might want to add an assert here to make sure.

Reply:
Why asserting ? I do not think it is semantically wrong to use `unknownLen()` in cases where the length could actually be deduced ( I understand it's not nice). I wanted to avoid special casing here but I will try to handle it if you think it is worth it.

Reply:
Well, the assert could be added if and only if the idea is that the LEN can never be constant here.

On the other hand, if the LEN can legitimately be constant, then this code should check for that and put the constant value in the type.

Lowering has been a little too cavalier in just throwing away this sort of information when it is not necessary to do so. It's not unequivocally wrong, no. But it is inferior.